org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java

```diff
@@ -476,33 +482,74 @@ public static Range selectPreviousNodesEndTag(int offset, DOMDocument document)
 	// ------------ Entities selection
 
 	/**
-	 * Returns the range of the used entity in a text node (ex : &amp;) and null
-	 * otherwise.
+	 * Returns the range of the entity reference in a text node (ex : &amp;) and
+	 * null otherwise.
 	 * 
 	 * @param offset   the offset
 	 * @param document the document
-	 * @return the range of the used entity in a text node (ex : &amp;) and null
-	 *         otherwise.
+	 * @return the range of the entity reference in a text node (ex : &amp;) and
+	 *         null otherwise.
 	 */
-	public static Range selectEntity(int offset, DOMDocument document) {
+	public static Range selectEntityReference(int offset, DOMDocument document) {
 		String text = document.getText();
 		// Search '&' character on the left of the offset
-		int startEntityOffset = offset - 1;
-		while (startEntityOffset > -1 && Character.isLetterOrDigit(text.charAt(startEntityOffset))) {
-			startEntityOffset--;
-		}
-		if (startEntityOffset == -1 || text.charAt(startEntityOffset) != '&') {
+		int entityReferenceStart = getEntityReferenceStartOffset(text, offset);
+		if (entityReferenceStart == -1) {
 			return null;
 		}
 		// Search ';' (or character on the right of the offset
-		int endEntityOffset = offset;
-		while (endEntityOffset < text.length() && Character.isLetterOrDigit(text.charAt(endEntityOffset))) {
-			endEntityOffset++;
+		int entityReferenceEnd = getEntityReferenceEndOffset(text, offset);
+		return createRange(entityReferenceStart, entityReferenceEnd, document);
+	}
+
+	/**
+	 * Returns the start offset of the entity reference (ex : &am|p;) from the given
+	 * offset and -1 if no entity reference.
+	 * 
+	 * @param text   the XML content.
+	 * @param offset the offset.
+	 * @return the start offset of the entity reference (ex : &am|p;) from the given
+	 *         offset and -1 if no entity reference.
+	 */
+	public static int getEntityReferenceStartOffset(String text, int offset) {
+		offset--;
+		if (offset < 0) {
+			return -1;
+		}
+		if (text.charAt(offset) == '&') {
+			return offset;
+		}
+		if (offset == 0) {
+			return -1;
+		}
+		int startEntityOffset = StringUtils.findStartWord(text, offset, ENTITY_NAME_PREDICATE);
+		if (startEntityOffset == -1) {
+			return -1;
+		}
+		if (startEntityOffset != -1 && text.charAt(startEntityOffset - 1) != '&') {

```

Comment:
Can `startEntityOffset == 0` here? In that case, I'm concerned about a possible IndexOutOfBoundsException here

Reply:
Good catch! I have fixed the problem and write several tests with this method (see XMLPositionUtilityTest#entityReference()).